workloads/dataset-controller/dataset-controller.py

```diff
@@ -0,0 +1,52 @@
+import os
+
+from kubernetes import config, client, watch
+from prometheus_client import Counter, start_http_server
+
+from ocifs import OCIFileSystem
+
+import pyarrow.feather
+from datasets import load_dataset
+
+DOMAIN = "new.artificialwisdom.cloud"
+RESOURCE_VERSION = ""
+
+MAIN_LOOP_ITERATIONS = Counter("main_loop_iterations",
+                               "Number of times main loop has restarted")
+OBJECTS_PROCESSED = Counter("objects_processed",
+                            "Number of objects processed")
+
+fs = OCIFileSystem(config=os.environ["OCI_CONFIG"], profile="DEFAULT")
+
+if "KUBERNETES_PORT" in os.environ:
+    config.load_incluster_config()
+else:
+    config.load_kube_config()
+
+start_http_server(8000)
+
+crds = client.CustomObjectsApi()
+while True:
+    MAIN_LOOP_ITERATIONS.inc()
+    stream = watch.Watch().stream(
+            crds.list_cluster_custom_object,
+            DOMAIN, "v1",
+            "datasets", resource_version=RESOURCE_VERSION)
+    for event in stream:
+        OBJECTS_PROCESSED.inc()
+        obj = event["object"]
+        source = obj["spec"]["source"]
+        if source.startswith("hf:"):
+            name = source[3:]
+            print("Retrieving dataset from HF", name)
+            ds = load_dataset(name)["train"]
```

Comment:
I believe datasets are easily batchable with a flag.

Reply:
Yeah. I'm not sure whether we can also parallelize writing to Feather files in OCI; I'd like to hope that it's possible and easy, but I didn't see anything like that in anybody's API docs.